### PR TITLE
Update sudoku.nim

### PR DESCRIPTION
**Pull request by flaviut**
Don't merge this, this is just to comment on the source.

### CODE REVIEW

File: sudoku.nim [hunk 1]
```diff
@@ -1,171 +1,171 @@
-import strutils
-
-type
-  TSudoku = array[0..80, 0..9] # 0 is blank
```
Comment (flaviut) [inline sudoku.nim:4]: This could be `array[9, array[9, 0..9]]`, which could be indexed as `sudoku[1][2]`

File: sudoku.nim [hunk 2]
```diff
@@ -1,171 +1,171 @@
...
-  else:
-    echo("Error clearing ", index, " of ", value)
-    printSudoku(sudoku)
-    echo blockPos[s.b]
-    echo colPos[s.c]
-    echo rowPos[s.r]
-    echo (blockPos[s.b] + colPos[s.c] + rowPos[s.r])
-    raise newException(EOS, "clear fail")
-
-proc loadSudoku(fileName: String): TSudoku =
```
Comment (flaviut) [inline sudoku.nim:109]: `String` should be lowercase, or it won't compile in the next version.

File: sudoku.nim [hunk 3]
```diff
@@ -1,171 +1,171 @@
-import strutils
-
-type
```
Comment (flaviut) [inline sudoku.nim:3]: The next version of Nim won't have `TFoo`s anymore, just `Foo`. Not a big deal though.

File: sudoku.nim [hunk 4]
```diff
@@ -1,171 +1,171 @@
...
-    raise newException(EOS, "clear fail")
-
-proc loadSudoku(fileName: String): TSudoku =
-  var file = open(fileName)
-  try:
-    var index = 0
-    while true:
-      let c = readChar(file)
-      try:
-        let n = parseInt($c) # There must be a better way than "parseInt($c)"
```
Comment (flaviut) [inline sudoku.nim:116]:
```
if c in {'0'..'9'}:
    let n = int(c - '0')
```

Not necessarily a better way, but a different way.

File: sudoku.nim [hunk 5]
```diff
@@ -1,171 +1,171 @@
...
-var # It would be better to pass all of these in to solve() as pointers, but
-    # I'm not sure how.
-  sudoku: TSudoku
-  blockPos: TSegmentPosList = getInitialSegmentPos()
-  colPos: TSegmentPosList = getInitialSegmentPos()
-  rowPos: TSegmentPosList = getInitialSegmentPos()
-
-# Utility functions
-
-proc printSudoku(sudoku: TSudoku) =
```
Comment (flaviut) [inline sudoku.nim:65]: better design might be to rename this to ``$`` and return a string. More composable and you can do `echo sudoku` if you want to print it.

File: sudoku.nim [hunk 6]
```diff
@@ -1,171 +1,171 @@
...
-    var str = ""
-    for y in 0..8:
-      if (y mod 3) == 0:
-        str = str & " "
-      str = str & " " & $sudoku[9 * x + y]
-    echo str
-    if (x mod 3) == 2:
-      echo ""
-
-proc setLoc(index, value: int) =
```
Comment (flaviut) [inline sudoku.nim:77]: See properties: http://nimrod-lang.org/manual.html#properties

File: sudoku.nim [hunk 7]
```diff
@@ -1,171 +1,171 @@
...
-  let s = getSegments(index)
-  if value in (blockPos[s.b] * colPos[s.c] * rowPos[s.r]):
-    sudoku[index] = value
-    excl(blockPos[s.b], value)
-    excl(colPos[s.c], value)
-    excl(rowPos[s.r], value)
-  else:
-    echo("Error setting ", index, " to ", value)
```
Comment (flaviut) [inline sudoku.nim:85]: It might be a better idea to put all this in the exception message rather than explicitly printing it.

File: sudoku.nim [hunk 8]
```diff
@@ -1,171 +1,171 @@
...
-# Create Variables
-
-proc getInitialSegmentPos(): TSegmentPosList =
-  var full: TPossibles
-  for i in 1..9:
-    incl(full, i)
-  for i in low(result)..high(result):
-    result[i] = full
-
-var # It would be better to pass all of these in to solve() as pointers, but
```
Comment (flaviut) [inline sudoku.nim:56]: Its possible, just do `proc foo(sudoku: var TSudoku)`

File: sudoku.nim (hunk 5 above)
Comment (aclough) [inline sudoku.nim:65]: Good idea

File: sudoku.nim (hunk 8 above)
Comment (aclough) [inline sudoku.nim:56]: Ah, I hadn't been thinking in those terms.

File: sudoku.nim (hunk 3 above)
Comment (aclough) [inline sudoku.nim:3]: Will there be type sensitivity?  In most languages I would use Foo for the type and foo for the instance.

File: sudoku.nim (hunk 1 above)
Comment (aclough) [inline sudoku.nim:4]: That didn't seem to really buy me anything here.

File: sudoku.nim (hunk 2 above)
Comment (aclough) [inline sudoku.nim:109]: I guess things are becoming case sensitive then?

File: sudoku.nim (hunk 4 above)
Comment (aclough) [inline sudoku.nim:116]: I was trying to figure out how to do that but missing the int( ... ) part.  That's the normal C idiom I tried to use first.

EDIT: Nevermind, when I try to use 'c - '0'' it says that - can't apply to two chars.

Comment (flaviut) [inline sudoku.nim:116]: Right, my bad. `int(c) - int('0')` works, you have to convert before performing arithmetic.

File: sudoku.nim (hunk 3 above)
Comment (flaviut) [inline sudoku.nim:3]: Yes, but only on the first character. `Foo` = `FoO`; `foo` ≠ `Foo`

**Pull request closed without merging**